stripe/src/main/java/com/stripe/android/view/CardNumberEditText.java

```diff
@@ -0,0 +1,175 @@
+        super(context, attrs);
+        listenForTextChanges();
+    }
+
+    public CardNumberEditText(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+        listenForTextChanges();
+    }
+
+    /**
+     * Updates the selection index based on the current (pre-edit) index, and
+     * the size change of the number being input.
+     *
+     * @param oldPosition the original position of the cursor
+     * @param oldLength the pre-edit length of the string
+     * @param newLength the post-edit length of the string
+     * @return an index within the string at which to put the cursor
+     */
+    @VisibleForTesting
+    int updateSelectionIndex(int oldPosition, int oldLength, int newLength) {
+        int increment, newPosition;
+        Set<Integer> spaceSet = Card.AMERICAN_EXPRESS.equals(mCardBrand)
+                ? SPACE_SET_AMEX
+                : SPACE_SET_COMMON;
+
+        boolean growing = newLength > oldLength;
+
+        if (growing) {
+            increment = spaceSet.contains(oldPosition) ? 2 : 1;
+            newPosition = oldPosition + increment;
+            return newPosition <= newLength ? newPosition : newLength;
+        } else {
+            increment = spaceSet.contains(oldPosition - 2) ? 2 : 1;
+            newPosition = oldPosition - increment;
+            return newPosition > 0 ? newPosition : 0;
+        }
+    }
+
+    private void listenForTextChanges() {
+        addTextChangedListener(new TextWatcher() {
+            boolean ignoreChanges = false;
+            int beforeStringLength = 0;
+            int beforeSelectionIndex = 0;
+
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                if (!ignoreChanges) {
+                    beforeStringLength = getText().toString().length();
+                    beforeSelectionIndex = getSelectionEnd();
+                }
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if (ignoreChanges) {
+                    return;
+                }
+
+                if (start < 4) {
+                    updateCardBrand(s.toString());
+                    // update the card icon here
+                }
+
+                if (beforeSelectionIndex > 16) {
+                    // no need to do formatting if we're past all of the spaces.
+                    return;
+                }
+
+                String spacelessNumber = StripeTextUtils.convertToSpacelessNumber(s.toString());
+                if (spacelessNumber == null) {
+                    return;
+                }
+
+                String[] cardParts = StripeTextUtils.separateCardNumberGroups(
+                        spacelessNumber, mCardBrand);
+                StringBuilder formattedNumberBuilder = new StringBuilder();
+                for (int i = 0; i < cardParts.length; i++) {
```

Comment:
Not a strong reason -- the last time I added a TextUtils function, we had a user complain because it added a library that he didn't use. I reverted the change that added it (I think I was just checking `TextUtils.isEmpty`), but I did specifically say that I couldn't promise not to use it in the future.